database-commons/src/main/java/io/cdap/plugin/db/connector/AbstractDBSpecificConnector.java

```diff
@@ -0,0 +1,124 @@
+
+/**
+ * An Abstract DB Specific Connector those specific DB connectors can inherits
+ * @param <T> the Record type that specific DB Record Reader may return while sample the data with InputFormat
+ */
+public abstract class AbstractDBSpecificConnector<T extends DBWritable> extends AbstractDBConnector
+  implements BatchConnector<LongWritable, T> {
+
+  protected final AbstractDBSpecificConnectorConfig config;
+
+  protected AbstractDBSpecificConnector(AbstractDBSpecificConnectorConfig config) {
+    super(config);
+    this.config = config;
+  }
+
+  public abstract boolean supportSchema();
+
+  protected abstract Class<? extends DBWritable> getDBRecordType();
+
+  protected SchemaReader getSchemaReader() {
+    return new CommonSchemaReader();
+  }
+
+  @Override
+  protected DBConnectorPath getDBConnectorPath(String path) throws IOException {
+    return DBSpecificPath.of(path, supportSchema());
+  }
+
+  @Override
+  public InputFormatProvider getInputFormatProvider(ConnectorContext context, SampleRequest request)
+    throws IOException {
+    DBSpecificPath path = DBSpecificPath.of(request.getPath(), supportSchema());
+    if (path.getTable() == null) {
+      throw new IllegalArgumentException(
+        String.format("Path %s cannot be sampled. Must have table name in the path.", request.getPath()));
+    }
+
+    ConnectionConfigAccessor connectionConfigAccessor = new ConnectionConfigAccessor();
+    if (config.getUser() == null && config.getPassword() == null) {
+      DBConfiguration
+        .configureDB(connectionConfigAccessor.getConfiguration(), driverClass.getName(),
+                     getConnectionString(path.getDatabase()));
+    } else {
+      DBConfiguration
+        .configureDB(connectionConfigAccessor.getConfiguration(), driverClass.getName(),
+                     getConnectionString(path.getDatabase()),
+                     config.getPassword(), config.getPassword());
+    }
+    String tableQuery = getTableQuery(path);
+    DataDrivenETLDBInputFormat.setInput(connectionConfigAccessor.getConfiguration(), getDBRecordType(),
+                                        tableQuery, null, false);
+    connectionConfigAccessor.setConnectionArguments(Maps.fromProperties(config.getConnectionArgumentsProperties()));
+    connectionConfigAccessor.getConfiguration().setInt(MRJobConfig.NUM_MAPS, 1);
+    try {
+      connectionConfigAccessor.setSchema(loadTableSchema(getConnection(path),  tableQuery).toString());
+    } catch (SQLException e) {
+      throw new IOException(String.format("Failed to get table schema due to: %s.", e.getMessage()), e);
+    }
+
+
+    return new SourceInputFormatProvider(DataDrivenETLDBInputFormat.class, connectionConfigAccessor.getConfiguration());
+  }
+
+  protected Connection getConnection(DBConnectorPath path) {
+    return getConnection(getConnectionString(path.getDatabase()) , config.getConnectionArgumentsProperties());
+  }
+
+  protected String getConnectionString(String database) {
```

Comment:
Is this overriden anywhere? I don't see the parameter used

Reply:
Yes, it's overridden in the `PostgresConnector`, because Postgres doesn't support  cross DB connection, need to put database in connection string.